Begin to purge empty names from the codebase

mitchellwrosen opened the pull request:
## Overview

This PR eliminates the `Monoid` (and an unused `Semigroup`) instance of `Name`, as a first step towards changing `Name`'s representation to address #2465.

The proposed representation does not have a sensible "empty" name, like the current `mempty` / `""` - which is a good thing, because we don't want empty `Name`s in the codebase anyway!  


## Implementation notes

The `Monoid` instance was only used in one place, in some search code, which constructed search results that contain definitely-named things (`HashQualified' Name`). Those have been refactored to `HashQualified Name` instead, which allows the `HashOnly` constructor to survive all the way through.

The refactoring is semantics-preserving, because the total conversion from a `HashQualified Name` to a `HashQualified' Name` that was being performed prior to this PR lives on, only now, we don't translate the `HashOnly` constructor to a `HashQualified` constructor with an empty name.

## Test coverage

No new tests were added, but I performed a manual test with the UI, and searching appears to work. That is what I would expect from the changes, as they are really just about propagating the transformation from

```haskell
HashQualified "" shorthash  -- bogus empty name
```

back to

```haskell
HashOnly shorthash
```

throughout the codebase.

## Loose ends

I don't think this PR eliminates all empty names from the codebase - `Name ""` may be lurking in various places. That will require some more spelunking.

mitchellwrosen committed:
Remove Monoid instance of Name

In doing so, the term/type names inside a SearchResult had to change
from HashQualified' (named) to HashQualified (possibly hash only), which
they effectively already were, albeit with bogus empty names in the
hash-only case.

mitchellwrosen committed:
Remove unused instance Semigroup Name

pchiusano reviewed (approved):
👍

aryairani commented:
I skimmed this but wasn't sure I was understanding the greater ramifications of the changes, if any.  Maybe we can run through it together on Tuesday if you want.

ChrisPenner commented on parser-typechecker/src/Unison/Server/Backend.hs:
```diff
@@ -468,47 +468,45 @@ fixupNamesRelative root = Names3.map0 fixName where
-            let hqNames = Names3.typeName len r names
+            let prioritize :: Set (HQ'.HashQualified Name) -> (HQ'.HashQualified Name, Set (HQ'.HashQualified Name))
+                prioritize =
+                  unsafeUnconsSet . sortOn (\n -> HQ.matchesNamedReference (HQ'.toName n) r query) . Set.toList
```
Seems possible that there isn't a valid name for our ref in our names here;
AFAICT with the use of `unsafeUnconsSet` it will error in that case;

I might be missing something, but it seems like it should probably return an empty list of matches for this query instead.

mitchellwrosen replied:
Hmm... I don't think it's possible for this code to pass an empty set into `prioritize` because we are looking up references that were just returned by `Names3.lookupRelativeHQType`.

For example, we might first call that on name `"foo.bar"`...

```haskell
> Names3.lookupRelativeHQType "foo.bar" names
{ "ref#1", "ref#2", "ref#3" }
```

...then look up, for each ref, all of its full names (including back in history)...

```haskell
> Names3.typeName len "ref#1" names
{ "oink.foo.bar", "blah.foo.bar#old" }
```

which is the set that we `prioritize` into a "preferred" name and a list of aliases.

Previously, the possibly-empty set here was handled by a `lastMay` and `mayName`, which was confusing, and introduced a bogus empty-name thing, too.

I could be missing something, but that's my understanding based on the code so far.

ChrisPenner commented on parser-typechecker/src/Unison/Server/Backend.hs:
```diff
@@ -468,47 +468,45 @@ fixupNamesRelative root = Names3.map0 fixName where
-                    in  SR.termResult primaryName r aliases
+            let prioritize :: Set (HQ'.HashQualified Name) -> (HQ'.HashQualified Name, Set (HQ'.HashQualified Name))
+                prioritize =
+                  unsafeUnconsSet . sortOn (\n -> HQ.matchesNamedReferent (HQ'.toName n) r query) . Set.toList
```
Same concern here; I think this should result in an empty list of matches for this query, not an error.

ChrisPenner commented on parser-typechecker/src/Unison/Server/Backend.hs:
```diff
@@ -468,47 +468,45 @@ fixupNamesRelative root = Names3.map0 fixName where
+    -- Uncons a non-empty list, and return the tail as a set. Calls 'error' if the given list is empty.
+    unsafeUnconsSet :: Ord a => [a] -> (a, Set a)
+    unsafeUnconsSet = \case
+      [] -> error "unconsSet: empty list"
```
Is there any way we can fix the types to prevent this instead?

If not, I think `unsafeUnconsSet` should accept an error message so we can explain what's happening and so it's easier to track down exactly what cause the error rather than a generic `unconsSet: empty list"`.

aryairani replied:
I like this idea.

aryairani commented on parser-typechecker/src/Unison/Codebase/Editor/HandleInput.hs:
```diff
@@ -1293,14 +1293,14 @@ loop = do
+                       <> (SR.typeName <$> typeResults fromHits)
             termToMisses = toMisses'
-                       <> (HQ'.toHQ . SR.typeName <$> typeResults toHits)
+                       <> (SR.typeName <$> typeResults toHits)
             -- Term hits are type misses
             typeFromMisses = fromMisses'
-                       <> (HQ'.toHQ . SR.termName <$> termResults fromHits)
+                       <> (SR.termName <$> termResults fromHits)
             typeToMisses = toMisses'
-                       <> (HQ'.toHQ . SR.termName <$> termResults toHits)
+                       <> (SR.termName <$> termResults toHits)
```
Here I was going to suggest hitting the region with ormolu.

aryairani commented on parser-typechecker/src/Unison/Codebase/Editor/HandleInput.hs:
```diff
@@ -2105,9 +2105,9 @@ listBranch (Branch.toNames0 -> b) =
-  _ -> error "unpossible match failure"
+  SR.Tm' n r _ -> HQ.toString $ HQ.requalify n r
+  SR.Tp' n r _ -> HQ.toString $ HQ.requalify n (Referent.Ref r)
+  _ -> error "impossible match failure"
```
This Simpsons reference has seen its last days.

mitchellwrosen replied:
Whoops I ruined someone's fun by not getting the reference

mitchellwrosen committed:
simplify search implementation a bit with a helper type

mitchellwrosen committed:
cleanup/refactoring

mitchellwrosen committed:
make search result aliases definitely have names, again

mitchellwrosen committed:
simplify searchBranchExact a bit

mitchellwrosen committed:
better error message in impossible case

mitchellwrosen committed:
inline searchBranchExact

mitchellwrosen commented:
This PR has changed a bit since reviewing yesterday with @aryairani and @ChrisPenner, please have a look over the new commits.

While addressing feedback, it occurred to me that the `SearchResult`'s aliases would never be nameless, so I reverted that back to `HQ'.HashQualified` as it was originally.

The primary name is still a `HQ.HashQualified`, per the changes in this PR that remove the `instance Monoid Name`.

Additionally, I did a bit of cleanup to reduce duplication in a couple places, where we had identical code, module `Reference`/`Referent` and associated helpers.

mitchellwrosen committed:
⅄ trunk → 21-10-07-no-empty-names
